### PR TITLE
ADM-480[Frontend]Feat: Update the select menu to autocomplete

### PR DESCRIPTION
**Pull request by 7758521yyy**
## Summary

...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: frontend/__tests__/src/components/Metrics/MetricsStep/CycleTime.test.tsx [hunk 1]
```diff
@@ -96,35 +96,100 @@ describe('CycleTime', () => {
...
+      await userEvent.click(columnsArray[2])
+      const listBox = within(getByRole('listbox'))
+      const options = listBox.getAllByRole('option')
+      // 找到被选中的选项
+      const selectedOption = options.find((option) => option.getAttribute('aria-selected') === 'true')
+
+      // 获取选中的选项文本
+      const selectedOptionText = selectedOption?.textContent
+
+      // 验证选中的选项是否为 "----"
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/CycleTime.test.tsx:139]: Same

File: frontend/__tests__/src/components/Metrics/MetricsStep/DeploymentFrequencySettings/PipelineMetricSelection.test.tsx [hunk 2]
```diff
@@ -149,13 +149,13 @@ describe('PipelineMetricSelection', () => {
...
     metricsClient.getSteps = jest.fn().mockReturnValue({ response: [], haveStep: false })
-    const { getByText, getByRole } = await setup(
+    const { getByText, getByRole, getAllByRole } = await setup(
       { id: 0, organization: 'mockOrgName', pipelineName: 'mockName', step: '', branches: [] },
       false,
       false
     )
 
-    await userEvent.click(getByRole('button', { name: PIPELINE_NAME }))
+    await userEvent.click(getAllByRole('button', { name: 'Open' })[1])
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/DeploymentFrequencySettings/PipelineMetricSelection.test.tsx:158]: All operation like click should be warped by Act method

**Review by guzhongren — COMMENTED** [review]
Change

File: frontend/__tests__/src/components/Metrics/MetricsStep/DeploymentFrequencySettings/PipelineMetricSelection.test.tsx (hunk 2 above)
Comment (7758521yyy) [reply frontend/__tests__/src/components/Metrics/MetricsStep/DeploymentFrequencySettings/PipelineMetricSelection.test.tsx:158]: > All operation like click should be warped by Act method

Thank you for reminding!
Seems like testing-library/react and userEvent already helped us to do this.
https://testing-library.com/docs/preact-testing-library/api/#act

File: frontend/__tests__/src/components/Metrics/MetricsStep/Crews.test.tsx [hunk 3]
```diff
@@ -39,57 +39,73 @@ describe('Crew', () => {
...
+    expect(getByRole('button', { name: 'crew A' })).toBeInTheDocument()
+    expect(getByRole('button', { name: 'crew B' })).toBeInTheDocument()
   })
 
   it('should show detail options when click Included crews button', async () => {
     const { getByRole } = setup()
-    await userEvent.click(getByRole('button', { name: mockLabel }))
+    await act(async () => {
+      await userEvent.click(getByRole('combobox', { name: mockLabel }))
+    })
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/Crews.test.tsx:52]: give `new line` for `given`, `when` and `then`

File: frontend/__tests__/src/components/Metrics/MetricsStep/CycleTime.test.tsx [hunk 4]
```diff
@@ -72,21 +68,27 @@ describe('CycleTime', () => {
...
+        await userEvent.click(columnsArray[0])
+      })
       const listBox = within(getByRole('listbox'))
       const options = listBox.getAllByRole('option')
-      const optionValue = options.map((li) => li.getAttribute('data-value'))
+      const optionText = options.map((option) => option.textContent)
 
-      expect(optionValue).toEqual([
+      const expectedOptions = [
         '----',
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/CycleTime.test.tsx:92]: those values should be defined in a fixture, and update the previous test case

File: frontend/__tests__/src/components/Metrics/MetricsStep/CycleTime.test.tsx [hunk 5]
```diff
@@ -96,35 +98,119 @@ describe('CycleTime', () => {
...
       const listBox = within(getByRole('listbox'))
-      const mockOption = listBox.getByRole('option', { name: 'To do' })
+      const options = listBox.getAllByRole('option')
+      const selectedOption = options.find((option) => option.getAttribute('aria-selected') === 'true')
 
-      await userEvent.click(mockOption)
+      const selectedOptionText = selectedOption?.textContent
 
-      expect(getByText('To do')).toBeInTheDocument()
+      expect(selectedOptionText).toBe('----')
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/CycleTime.test.tsx:146]: to be a fixture

File: frontend/__tests__/src/components/Metrics/MetricsStep/DeploymentFrequencySettings/PipelineMetricSelection.test.tsx [hunk 6]
```diff
@@ -170,15 +180,19 @@ describe('PipelineMetricSelection', () => {
...
   it('should show no steps warning message when getSteps succeed but get no steps and isShowRemoveButton is true', async () => {
     metricsClient.getSteps = jest.fn().mockReturnValue({ response: [], haveStep: false })
-    const { getByRole } = await setup(
+    const { getByRole, getAllByRole } = await setup(
       { id: 0, organization: 'mockOrgName', pipelineName: 'mockName', step: '', branches: [] },
       true,
       false
     )
+    await act(async () => {
+      await userEvent.click(getAllByRole('button', { name: 'Open' })[1])
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/DeploymentFrequencySettings/PipelineMetricSelection.test.tsx:189]: `Open` should be a fixture

File: frontend/__tests__/src/components/Metrics/MetricsStep/DeploymentFrequencySettings/SingleSelection.test.tsx [hunk 7]
```diff
@@ -1,4 +1,4 @@
-import { render } from '@testing-library/react'
+import { act, getByText, render, within } from '@testing-library/react'
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/DeploymentFrequencySettings/SingleSelection.test.tsx:1]: remove `getByText,`

File: frontend/__tests__/src/components/Metrics/MetricsStep/Crews.test.tsx (hunk 3 above)
Comment (7758521yyy) [reply frontend/__tests__/src/components/Metrics/MetricsStep/Crews.test.tsx:52]: > give `new line` for `given`, `when` and `then`

Nice catch

File: frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        isError={isError}
+        testId={testId}
+      />
+    )
+
+    const autoCompleteComponent = screen.getByTestId(testId)
+    expect(autoCompleteComponent).toBeInTheDocument()
+  })
+
+  it('When passed selectedoption changed, the correct option would be displayed', async () => {
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx:33]: use `should ... when... given..` pattern

File: frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      <MultiAutoComplete
+        optionList={optionList}
+        selectedOption={selectedOption}
+        onChangeHandler={onChangeHandler}
+        isSelectAll={isSelectAll}
+        textFieldLabel={textFieldLabel}
+        isError={isError}
+        testId={testId}
+      />
+    )
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx:44]: extract it into a `setup` method

File: frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    expect(screen.getByRole('button', { name: 'Option 1' })).toBeVisible()
+  })
+
+  it('When All, All would be called by onChange function', async () => {
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx:49]: Update test name

File: frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      <MultiAutoComplete
+        optionList={optionList}
+        selectedOption={selectedOption}
+        onChangeHandler={onChangeHandler}
+        isSelectAll={isSelectAll}
+        textFieldLabel={textFieldLabel}
+        isError={isError}
+        testId={testId}
+      />
+    )
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx:60]: refactor

File: frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    await userEvent.click(inputField)
+
+    const allOption = screen.getByRole('option', { name: 'All' })
+    await act(async () => {
+      await userEvent.click(allOption)
+    })
+
+    expect(onChangeHandler).toHaveBeenCalledWith(expect.anything(), ['Option 1', 'All'], 'selectOption', {
+      option: 'All',
+    })
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/MetricsStep/MultiAutoComplete.test.tsx:72]: extract those into fixtures

File: frontend/cypress/e2e/createANewProject.cy.ts [hunk 13]
```diff
@@ -185,6 +187,18 @@ const checkFieldsExist = (fields: string[]) => {
   })
 }
 
+const checkAutoCompleteFieldsExist = (fields: { name: string; value: string }[]) => {
+  fields.forEach((item) => {
+    cy.contains(item?.name)
+      // 获取同级的 input 元素
```
Comment (guzhongren) [inline frontend/cypress/e2e/createANewProject.cy.ts:193]: remove comments

File: frontend/cypress/e2e/createANewProject.cy.ts [hunk 14]
```diff
@@ -185,6 +187,18 @@ const checkFieldsExist = (fields: string[]) => {
...
+      .siblings()
+      .eq(0)
+      .find('input')
+      // 验证输入框的值
```
Comment (guzhongren) [inline frontend/cypress/e2e/createANewProject.cy.ts:197]: remove comments

File: frontend/cypress/e2e/importAProject.cy.ts [hunk 15]
```diff
@@ -56,6 +63,18 @@ const checkFieldsExist = (fields: string[]) => {
   })
 }
 
+const checkAutoCompleteFieldsExist = (fields: { name: string; value: string }[]) => {
+  fields.forEach((item) => {
+    cy.contains(item?.name)
+      // 获取同级的 input 元素
```
Comment (guzhongren) [inline frontend/cypress/e2e/importAProject.cy.ts:69]: remove comments

File: frontend/cypress/e2e/importAProject.cy.ts [hunk 16]
```diff
@@ -56,6 +63,18 @@ const checkFieldsExist = (fields: string[]) => {
...
+      .siblings()
+      .eq(0)
+      .find('input')
+      // 验证输入框的值
```
Comment (guzhongren) [inline frontend/cypress/e2e/importAProject.cy.ts:73]: remove comments

File: frontend/cypress/pages/metrics/metrics.ts [hunk 17]
```diff
@@ -1,83 +1,80 @@
 class Metrics {
   private readonly cycleTimeSettingAnalysis = () => {
     cy.contains('In Analysis').siblings().eq(0).click()
-    cy.get('[data-test-id="Analysis"]:contains(Analysis)').click()
+    cy.get('li[role="option"]').contains('Analysis').click()
```
Comment (guzhongren) [inline frontend/cypress/pages/metrics/metrics.ts:4]: extract it into a constant

File: frontend/src/components/Common/MultiAutoComplete/index.tsx [hunk 18]
```diff
@@ -0,0 +1,59 @@
+import { Checkbox, createFilterOptions, TextField } from '@mui/material'
+import React from 'react'
+import { StyledAutocompleted } from './styles'
+
+type Props = {
+  optionList: string[]
+  selectedOption: string[]
+  // There is an any because m-ui strictly define its type, but the parameters are not that strict. Maybe because of version diff
+  onChangeHandler: any
```
Comment (guzhongren) [inline frontend/src/components/Common/MultiAutoComplete/index.tsx:9]: not recommend to use `any`, just a function, you can search online

File: frontend/src/components/Common/MultiAutoComplete/index.tsx [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+      value={selectedOption}
+      filterOptions={(options, params) => {
+        const filtered = filter(options, params)
+        return ['All', ...filtered]
+      }}
+      getOptionLabel={(option) => option as string}
+      onChange={onChangeHandler}
+      renderOption={(props, option, { selected }) => {
+        const selectAllProps =
+          option === 'All' // To control the state of 'select-all' checkbox
```
Comment (guzhongren) [inline frontend/src/components/Common/MultiAutoComplete/index.tsx:41]: remove comments

File: frontend/src/components/Common/MultiAutoComplete/index.tsx [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+            ? { checked: isSelectAll }
+            : {}
+        return (
+          <li {...props}>
+            <Checkbox style={{ marginRight: 8 }} checked={selected} {...selectAllProps} />
+            {/*The same reason. m-ui declare its type to unknown, but actually it can/should be string here. Maybe caused by diff version*/}
```
Comment (guzhongren) [inline frontend/src/components/Common/MultiAutoComplete/index.tsx:47]: remove comments

File: frontend/src/components/Common/MultiAutoComplete/styles.ts [hunk 21]
```diff
@@ -0,0 +1,10 @@
+import { styled } from '@mui/material/styles'
+import { Autocomplete } from '@mui/material'
+
+export const StyledAutocompleted = styled(Autocomplete)`
+  & .MuiAutocomplete-tag {
+    background-color: transparent;
+    border: 1px solid rgba(0, 0, 0, 0.26);
+    font-size: 14px;
```
Comment (guzhongren) [inline frontend/src/components/Common/MultiAutoComplete/styles.ts:8]: should not use `px`

File: frontend/src/components/Metrics/MetricsStep/Crews/index.tsx [hunk 22]
```diff
@@ -1,84 +1,66 @@
...
 interface crewsProps {
   options: string[]
   title: string
   label: string
 }
 
+export const WarningMessage = styled('p')({
+  color: 'red',
+})
+
```
Comment (guzhongren) [inline frontend/src/components/Metrics/MetricsStep/Crews/index.tsx:20]: put this part into a separated file named `style.tsx`

File: frontend/src/components/Common/MultiAutoComplete/index.tsx (hunk 18 above)
Comment (7758521yyy) [reply frontend/src/components/Common/MultiAutoComplete/index.tsx:9]: > not recommend to use `any`, just a function, you can search online

While...I do searched it but if I use the function with parameters type that the mui offical website provided, the type system would complain about it. That's why I left that comments above.

Comment (7758521yyy) [reply frontend/src/components/Common/MultiAutoComplete/index.tsx:9]: > not recommend to use `any`, just a function, you can search online

I suspect it maybe  the dependency version issue?

File: frontend/src/components/Common/MultiAutoComplete/styles.ts (hunk 21 above)
Comment (7758521yyy) [reply frontend/src/components/Common/MultiAutoComplete/styles.ts:8]: > should not use `px`

What we were used?
`pt`? `rem`?

Comment (guzhongren) [reply frontend/src/components/Common/MultiAutoComplete/styles.ts]: rem

File: frontend/src/components/Common/MultiAutoComplete/styles.ts [hunk 23]
```diff
@@ -0,0 +1,10 @@
+import { styled } from '@mui/material/styles'
+import { Autocomplete } from '@mui/material'
+
+export const StyledAutocompleted = styled(Autocomplete)`
+  & .MuiAutocomplete-tag {
+    background-color: transparent;
+    border: 1px solid rgba(0, 0, 0, 0.26);
```
Comment (guzhongren) [inline frontend/src/components/Common/MultiAutoComplete/styles.ts:7]: use `rem`, not others

**Review by guzhongren — COMMENTED** [review]
update

**Review by guzhongren — APPROVED** [review]

**Pull request merged**